feat: add support for AEDI 8

alsami opened the pull request:
* update to A.E.D.I version `8.0.0`
* remove projects and TFMs for `netcoreapp2.1` because M.E.D.I does not support it anymore -> this is a breaking change
* currently removed TFMs for test-projects since there is a name conflict because `ServiceCollection` now exists in `Microsoft.Extensions.DependencyInjection.Abstractions` & `Microsoft.Extensions.DependencyInjection`

Todos:
- [x] `RequestServicesFeatureFactory` might not be needed anymore but if it is still needed it can be simplified because we are not able to support `netcoreapp2.1` with `Microsoft.DependencyInjection` version 6
- [x] Add specification test to see if the container conforms to the expected behavior

#43

Commit:
feat: add support for AEDI 8

alsami commented:
@tillig my initial assumption was wrong. The tenant specifics requests do not work anymore. Which obviously had to happen because `IServiceScopeFactory` is now a singleton and is not created per tenant anymore.

I will need to look into it a bit more and think of a possible work around - right now I don't have any ideas on how to fix this.

Commit:
feat: add potential workaround for multitenancy with ASP.NET Core

alsami commented:
I am not sure if this problem can be really solved. I added my own implementation of `IServiceScopeFactory` and copied `AutofacServiceScope` from A.E.D.I. I've registered the former as transient in the existing `AutofacMultitenantServiceProviderFactory` and everything works.

Now all the breaking changes happend because `IServiceScopeFactory` was supposed to be a singleton according to MS. With this workaround we would bypass that requirement again and folks may run into issues. 

Any thoughts on this?

tillig commented:
The short and unfortunate version is: no, I don't have any thoughts, at least not right now. I mentioned in the original issue over on the dotnet/runtime repo this was going to be a real challenge. I figured this wouldn't be a five minute fix.

That doesn't help, but... at least you know you're not alone? I won't have time to look at this for a bit. Maybe I'll have more when I can get out from underwater and spend some time here.

alsami commented:
Fair enough. Will think about it a bit as well and get back to it later.

alsami commented:
So I thought a bit about a solution that is viable and doesn't break the contract. What if we were manually composing `IServiceScopeFactory` once for each tenant and the root within `MultitenantRequestServicesMiddleware`?

It would be a bit hacky and also surely violate a few principals but I think this way we could make it work. I am not sure if there are cases where this would be a problem but since we always set `IServiceProvidersFeature` within the middleware, automatically resolved dependencies would always receive the same scope (root, tenant-a. b or whatever).

Only problem I could think of is cases where people would have hosted-services running in their ASP.NET Core app. AFAIK this wouldn't work with tenant-specific dependencies anyway unless they are directly accessing multi-tenant container.

I might be on the wrong track here so I am really happy for any input

tillig commented:
This past weekend was a holiday weekend in the US and I was out with my family; I'm stuck in production issues for work at the moment along with some family stuff going on. What I can offer is how I'd approach looking at the problem, but I can't speak to specifics because, sadly, I will likely not have time to really dig into actual code and testing for... probably at least a week?

I'd probably start by trying to figure out what it'd take to make the multitenant container pass the M.E.DI spec tests using A.E.DI after the tenant has been identified. I think removing the complexity of the ASP.NET Core part (middleware and request services and all that) will be important to breaking this down. My gut says each tenant is going to need their own service scope factory so the singleton test passes, which is sort of where it sounds like you were heading.

I get the feeling that this may mean we need to have a multitenant-aware A.E.DI. That _could be in here_, or it may be a separate library. I don't have a good thought around this because I haven't put time into thinking about all the use cases or ramifications of splitting that out. I don't even know if it's really required, just brainstorming.

Once we know what it'll take to get the M.E.DI spec tests passing, it'll be a lot clearer what needs to happen specifically with ASP.NET Core to support this. Maybe it means we don't actually have to change the middleware, or the changes would be different.

That's all I have at the moment. Sorry if that's not helpful. Perhaps @alistairjevans has more, but IIRC he's pretty swamped at the moment, too. I'm probably going to duck out of this PR for a while unless there's a very specific question I can answer. I don't mean to sound like I'm abandoning you here, I'll be back, but I can't really offer "general guidance or thoughts" as such because that actually takes a lot of time and research, you know? Diving into the code, actually figuring out all the options, documenting the pros and cons of each, etc. It's a lot of work and a lot of time I just... I just don't have _right now_. I was pretty open last week when I was rolling out the other changes, but a lot of stuff hit all at once and there's nothing I can do about that. 😦

tillig commented on src/Autofac.Integration.AspNetCore.Multitenant/AutofacServiceScope.cs:
```diff
@@ -0,0 +1,81 @@
+/// Autofac implementation of the ASP.NET Core <see cref="IServiceScope"/>.
+/// </summary>
+/// <seealso cref="IServiceScope" />
+internal class AutofacServiceScope : IServiceScope, IAsyncDisposable
```
If it turns out we need to create service scopes directly, it might be better to make this public in the A.E.DI library.

But I'm wondering if we can make use of the [AutofacChildLifetimeScopeServiceProviderFactory](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/develop/src/Autofac.Extensions.DependencyInjection/AutofacChildLifetimeScopeServiceProviderFactory.cs) here, where each tenant has their own instance of this child scope provider factory. That looks a lot like what you've started building here.

tillig replied:
Hmmm. I may have spoken a little too soon. Let me work on a larger comment, but this is still along the lines of what I'm thinking.

tillig commented:
OK, I've pulled out from underwater for a bit and started digesting here. I think you're _generally_ on the right track.

I think we'll need to:

- Use `AutofacChildLifetimeScopeServiceProviderFactory` to our advantage.
  - One of these registered per tenant, in the tenant's scope (`InstancePerTenant`), similar to what we have [in `builder.Populate()`](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/c715e5f6c82f6481c299411fd0ec1f746f5f6aa5/src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs#L78).
  - "Root" for these should be the tenant scope.
- Root container should have a regular `AutofacServiceScopeProviderFactory` just like there already is.

We can probably register everything as `IServiceScopeFactory` but I'm concerned about `InstancePerTenant` [being basically `InstancePerMatchingLifetimeScope`](https://github.com/autofac/Autofac.Multitenant/blob/5d2523373bf980b496e42c5bff719b0ccaa59ecd/src/Autofac.Multitenant/RegistrationExtensions.cs#L36) - I suppose as long as no one tries to resolve `IEnumerable<IServiceScopeFactory>` [we should be OK.](https://github.com/autofac/Autofac/issues/969) I'm not sure resolving a set of those would be a real use case, though.

We might be able to do the `InstancePerTenant` registration right in the `AutofacMultitenantServiceProviderFactory` automatically.

Anyway, that's what I'm thinking so far. Does that help?

tillig commented:
Oh, and there's [some crazy conditional stuff](https://github.com/autofac/Autofac.AspNetCore.Multitenant/blob/f5a276c027b05bb3c9a7a2ac992851e1e15abbbc/src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs#L11) that goes on to support `netstandard 2.0` and ASP.NET Core pre-2.1. If we're removing support for earlier ASP.NET versions, we might not have to do all the lambda expression composition in the `RequestServicesFeatureFactory`. We may not even _need that_ anymore.

alsami commented:
First of all, thanks for the feeback, appreciate it.

> I think we'll need to:
> 
> * Use `AutofacChildLifetimeScopeServiceProviderFactory` to our advantage.
>   
>   * One of these registered per tenant, in the tenant's scope (`InstancePerTenant`), similar to what we have [in `builder.Populate()`](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/c715e5f6c82f6481c299411fd0ec1f746f5f6aa5/src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs#L78).
>   * "Root" for these should be the tenant scope.
> * Root container should have a regular `AutofacServiceScopeProviderFactory` just like there already is.

I am not sure how I would register that so that we would get an instance of `IServiceScopeFactory` per tenant. `AutofacChildLifetimeScopeServiceProviderFactory` implements `IServiceProviderFactory<T>`, what we need is an `IServiceScopeFactory` that is retrieved once but for each tenant and the root-lifetime. Can you elaborate a little bit for me please?

What if we were to make `AutofacServiceScopeFactory` public or at least add the attribute for internals to be exposed to this library. Then we could register it with `InstancePerTenant`, which would then be one for the root and one for each actual tenant as far as I can tell?

> I suppose as long as no one tries to resolve `IEnumerable<IServiceScopeFactory>` [we should be OK.](https://github.com/autofac/Autofac/issues/969) I'm not sure resolving a set of those would be a real use case, though.

Yeah I thought about that too - I don't think that there is any valid use-case to resolve `IEnumerable<IServiceScopeFactory>`  but then again it feels a bit weird to have more than one registration for the same interface for this specific case.

> Oh, and there's [some crazy conditional stuff](https://github.com/autofac/Autofac.AspNetCore.Multitenant/blob/f5a276c027b05bb3c9a7a2ac992851e1e15abbbc/src/Autofac.Integration.AspNetCore.Multitenant/RequestServicesFeatureFactory.cs#L11) that goes on to support `netstandard 2.0` and ASP.NET Core pre-2.1. If we're removing support for earlier ASP.NET versions, we might not have to do all the lambda expression composition in the `RequestServicesFeatureFactory`. We may not even _need that_ anymore.

Yeah true, we probably don't need that anymore but I will need to investigate - haven't done that so far.

Commit:
feat: register AutofacMultitenantServiceScopeFactory as InstancePerTenant

tillig commented:
> `AutofacChildLifetimeScopeServiceProviderFactory` implements `IServiceProviderFactory<T>`, what we need is an `IServiceScopeFactory` that is retrieved once but for each tenant and the root-lifetime. Can you elaborate a little bit for me please?

It's hard to be super explicit without... well, without, honestly, just doing the work. It seemed like you were on it, so I didn't create a branch and start really jamming on it myself because... if I figured out how to get it to work myself, then that'd kinda _be the PR_. I'm trying to put concepts and things together across these two repos basically without just doing the work. So... the high level concept I was thinking about was how you can use the service scope factory to get `Populate` and all that built on a lifetime scope, right?

```c#
var builder = new ContainerBuilder();
var container = builder.Build();
var fakeTenantScope = container.BeginLifetimeScope();
var providerFactory = new AutofacChildLifetimeScopeServiceProviderFactory(fakeTenantScope);
var services = new ServiceCollection();
var configAdapter = providerFactory.CreateBuilder(services);
var tenantServiceProvider = providerFactory.CreateServiceProvider(configAdapter);

// Here's a service scope factory that's a singleton for a lifetime scope.
var tenantServiceScopeFactory = tenantServiceProvider.GetService<IServiceScopeFactory>();
```

That is, the factory has the stuff to register all the things that are internal or whatever, so maybe that's something that could be used? However... maybe not? Maybe that won't work? Is there something there? Maybe?

As you noticed, I did start getting a little confused between lifetime scopes and service scope factories and Autofac scopes and scope providers and scope scope scopey-scopes and I'm sorry if that got confusing. Maybe what I mentioned earlier doesn't make sense.

Let me know if it turns out this isn't something you think you're going to be able to smash into existence and I'll see if I can create a branch and figure it out. It may be that we don't have everything public in the A.E.DI library that needs to be. I honestly don't know. I don't think I'd want to do an InternalsVisibleTo because, ideally, we're composing this thing in the same way someone _else_ would compose it, and if we don't have all the stuff exposed that we need to extend the features then maybe we need to fix that.

It'd still be worth getting the .NET 2.1 fixes in, removing any of the extra complexity that doesn't need to be there, so we could still pull that in when you get it sorted if you decide to not do the rest. Either way is fine, no problemo.

And again, sorry for the confusion. This is definitely some hard stuff.

alsami commented:
> It's hard to be super explicit without... well, without, honestly, just doing the work. It seemed like you were on it, so I didn't create a branch and start really jamming on it myself because... if I figured out how to get it to work myself, then that'd kinda _be the PR_. I'm trying to put concepts and things together across these two repos basically without just doing the work. So... the high level concept I was thinking about was how you can use the service scope factory to get `Populate` and all that built on a lifetime scope, right?

You are right and it is absolutely not my goal to make you work more than required. I am just lost here at the moment. It's rather challenging but I am certain that we'll find a solution.

> ```cs
> var builder = new ContainerBuilder();
> var container = builder.Build();
> var fakeTenantScope = container.BeginLifetimeScope();
> var providerFactory = new AutofacChildLifetimeScopeServiceProviderFactory(fakeTenantScope);
> var services = new ServiceCollection();
> var configAdapter = providerFactory.CreateBuilder(services);
> var tenantServiceProvider = providerFactory.CreateServiceProvider(configAdapter);
> 
> // Here's a service scope factory that's a singleton for a lifetime scope.
> var tenantServiceScopeFactory = tenantServiceProvider.GetService<IServiceScopeFactory>();
> ```

This is exactly what got me confused initially. We would create one more instance of `IServiceProvider`. For this specific package we already got a custom one which is `AutofacMultitenantServiceProviderFactory`. I remember that we had a bug because of two existing instances of `IServiceProvider` (#22 ) with the initial adjustments that were required for `netcoreapp3.0` and the breaking-changes that came with it.

> As you noticed, I did start getting a little confused between lifetime scopes and service scope factories and Autofac scopes and scope providers and scope scope scopey-scopes and I'm sorry if that got confusing. Maybe what I mentioned earlier doesn't make sense.

I feel you, still have the same problem every time I open up this solution😄 

I will need some more time investigating to figure out if we can really leverage the `AutofacChildLifetimeServiceProviderFactory`.

tillig commented:
No problem. Let me know if you get stuck enough that I should branch and work on it as well. I definitely think this new singleton scope provider thing is throwing a wrench in the works. Yay for conforming containers?

Commit:
feat: add delegate ServiceScopeFactory to compose the IServiceScopeFactory

alsami commented:
> No problem. Let me know if you get stuck enough that I should branch and work on it as well. I definitely think this new singleton scope provider thing is throwing a wrench in the works. Yay for conforming containers?

Wow. I am not sure if I should be happy or not but your suggestion worked. I've added an implementation with the latest commit. I had to also remove some older TFMs because of the name clash of `ServiceCollection`. 

Could you please review it and tell me If I am missing something with that solution?

**Edit**: Just noticed that while my solution works but it would call the implementation-factory actually more than once. I think I will somehow need to keep track of the existing factories that have been created already. Is there something I can use for that purpose?

tillig commented on test/Autofac.Integration.AspNetCore.Multitenant.Test/Autofac.Integration.AspNetCore.Multitenant.Test.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp2.1;netcoreapp3.1;net5.0;net6.0</TargetFrameworks>
+    <TargetFramework>net6.0</TargetFramework>
```
Check out the A.E.DI test project now. It still tests under all the TFMs, but there's NOT conditional references anymore. That fixes a lot of the compatibility problems. A.E.DI always requires M.E.DI 6+ now.

alsami replied:
I've re-added the other TFMs (netcoreapp3.1 and 5.0).

I still had to do the explicitly reference M.E.D.I. I think A.E.D.I is a bit different compared to this lib because we are also referencing ASP.NET Core including `Microsoft.AspNetCore.TestHost`.

alsami replied:
Okay nvm, I really was able to remove it from the test-project.

tillig commented on src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj:
```diff
@@ -47,14 +47,16 @@
 
   <ItemGroup>
-    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="7.2.0" />
+<!--    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="7.2.0" />-->
```
```suggestion
```

No need to keep the old one.

alsami replied:
Adressed with latest revision.

tillig commented on src/Autofac.Integration.AspNetCore.Multitenant/Autofac.Integration.AspNetCore.Multitenant.csproj:
```diff
@@ -47,14 +47,16 @@
     <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.6.1">
       <PrivateAssets>All</PrivateAssets>
     </PackageReference>
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="6.0.0" />
```
```suggestion
```

A.E.DI will already bring in M.E.DI 6, I think. Or, at least, the abstractions, which... is enough?

alsami replied:
There is a name clash because of `IServiceCollection` will see how to fix it.

alsami replied:
![image](https://user-images.githubusercontent.com/2729205/172413857-25d47c38-44d3-4966-a6f2-bf0ff1e3bc71.png)

The problem is that it has been part of `Microsoft.Extension.DependencyInjection.Abstractions` before `netcoreapp6.0`.

I am now conditionally including it so that lower versions retrieve the same package-reference.

tillig replied:
But if the .csproj stops doing conditional inclusions based on TFM, everything will be M.E.DI v6 and it'll just work... right?

alsami replied:
I tried that. I removed all conditional imports from the library project and I would still retrieve that error.

It is because of the framework-reference, which is different based on the TFM.

Basically this line --> `<FrameworkReference Include="Microsoft.AspNetCore.App" />`

tillig replied:
Can we... _not_ reference that? I mean, I guess I don't feel super strongly about the direct `<PackageReference>`, just thinking I don't want to get into a weird spot where there's some sort of transient dependency conflict. If you can't figure it out easily, I'm fine with leaving it.

tillig commented on src/Autofac.Integration.AspNetCore.Multitenant/AutofacMultitenantServiceProviderFactory.cs:
```diff
@@ -65,6 +66,17 @@ public IServiceProvider CreateServiceProvider(ContainerBuilder containerBuilder)
 
+            containerBuilder.Register<ServiceScopeFactory>(_ =>
+            {
+                return scope =>
```
NICE! Great work with this. Honestly just a tiny bit surprised that it works, but I'm not going to question success. 😆

tillig commented on src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs:
```diff
@@ -11,6 +11,13 @@
+    /// </summary>
+    /// <param name="lifetimeScope">The <see cref="ILifetimeScope"/> which can either be the root scope or dedicated tenant-scope.</param>
+    /// <returns>The <see cref="IServiceScopeFactory"/> for the specific <see cref="ILifetimeScope"/> which can either be the root scope or dedicated tenant-scope.</returns>
+    internal delegate IServiceScopeFactory ServiceScopeFactory(ILifetimeScope lifetimeScope);
```
I wonder if there's a better name for this since there are already so many of these overloaded "service scope factory provider accessor" sorts of things. I don't have suggestions, and I'm not going to block the PR over it, just thinking out loud. I mean, it's internal, so we can always change it later without breaking the API.

alsami replied:
Yeah it is kind of confusing. I can only think of really long names. 

What if we were to remove `Factory` and use the term `Accessor` here instead? Or `Activator`?

tillig replied:
I mean, technically it's _getting an `IServiceScopeFactory`_ so it's a `ServiceScopeFactoryAccessor`. Which isn't better, but is probably better than `ServiceScopeFactoryFactory`. We could go totally astray and call it `GetServiceScopeFactory` to explain what it _does_, which is... uh... it... gets service scope factories. I'm not in love with any of these. 😦

alsami replied:
I renamed it to `ServiceScopeFactoryAccessor` for now. We can change it at anytime since it is internal and can only be accessed by us (exception of reflection stuff).

tillig commented on src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs:
```diff
@@ -56,7 +70,8 @@ public async Task Invoke(HttpContext context)
             {
-                var autofacFeature = RequestServicesFeatureFactory.CreateFeature(context, _multitenantContainer.Resolve<IServiceScopeFactory>());
+                var factory = _serviceScopeFactory.Invoke(_multitenantContainer.GetCurrentTenantScope());
+                var autofacFeature = RequestServicesFeatureFactory.CreateFeature(context, factory);
```
It'd still be interesting to see if we even need this `RequestServicesFeatureFactory` now that there's no .NET 2.1.

alsami replied:
Would you want to remove the entire procedure to build the expression or only the bits with switching based on parameter count?

tillig replied:
Honestly, if we can skip the whole feature and just new-up the thing and have it work, let's do that. There's not a lot of value in having all this extra expression building to wrap a simple `new` operation.

alsami replied:
I pushed a simplified version that seems to be working targeting `netcoreapp3.1`, `net5.0`  & `net6.0`.

tillig reviewed (changes requested):
Really great work figuring this one out. What a tricky thing! Couple of pretty minor comments, but I think this will be ready quickly.

tillig commented:
Just thought of something that might be interesting to try: set up the M.E.DI compatibility tests but with a tenant container.

That is...

- [Add a reference to the M.E.DI spec tests](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/c715e5f6c82f6481c299411fd0ec1f746f5f6aa5/test/Autofac.Extensions.DependencyInjection.Test/Autofac.Extensions.DependencyInjection.Test.csproj#L41)
- [Create a `DependencyInjectionSpecificationTests` implementation](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/c715e5f6c82f6481c299411fd0ec1f746f5f6aa5/test/Autofac.Extensions.DependencyInjection.Test/Specification/BuilderSpecificationTests.cs#L9) where the way to build the `IServiceProvider` is:
  - Build the multitenant container.
  - Set a "known tenant" so the tenant is always identified.
  - Get the tenant's `IServiceProvider` and test against that.

That'd let us know if the logic here is going to hold. It may be interesting to have an internal static method that handles that little magic `AutofacChildLifetimeScopeServiceProviderFactory` registration so it can be used to wire that up in tests; or use the `AutofacMultitenantServiceProviderFactory` to build up the service provider so it's done the same as in the middleware.

Commit:
feat: keep track of the instances of IServiceScopeFactory that have already been created for a tenant-scope

Commit:
feat: include M.E.D.I directly but only if TFM is not .NET 6.0

Commit:
feat: re-add tests to be executed against .NET Core 3.1 and 5.0 as well

Commit:
chore: remove explicit reference for M.E.D.I. in test-project

tillig commented on src/Autofac.Integration.AspNetCore.Multitenant/MultitenantRequestServicesMiddleware.cs:
```diff
@@ -82,5 +115,29 @@ public async Task Invoke(HttpContext context)
             }
         }
+
+        private async Task<IServiceScopeFactory> GetServiceScopeFactoryForTenantLifetimeScope(CancellationToken httpContextRequestAborted)
```
Doesn't Autofac already handle locking around resolving singletons/instance-per-lifetime-scope? Are we putting a perf bottleneck in here that doesn't need to be here?

alsami replied:
Sadly we have to somehow manually handle that specific case here. I double checked, the implementation-factory is called several times for the same tenant, causing an `InstancePerDependency()` behavior even though I used `.SingleInstance()`. 

That's why I have added that caching mechanism here. Should the delegate really just be called once?

tillig replied:
What about `InstancePerTenant` (which is basically `InstancePerMatchingLifetimeScope`) instead of `SingleInstance`? Would that fix it?

alsami replied:
The first problem with that is, that the middleware directly injects the new `ServiceScopeFactory` delegate, which doesn't work since the first time it is being resolved, there is no tenant scope just the regular one. 

I tried working around that by manually locating the `ServiceScopeFactory` using the MTC. Sadly also then, the implementation-factory in `AutofacMultitenantServiceProviderFactory` is always called. It's odd to be honest, I'd have expected it to be resolved once.

alsami replied:
I agree that the solution is not good in any case. I think I will leave it for today and get back to it again with some fresh thoughts - also trying to implement the specification as you have suggested. I think that would help a lot. Maybe more problems will arise with that, hopefully not.

alsami replied:
I removed the manual locking and storing of cached instances of `IServiceProviderFactory` for each tenant. Sadly, the `InstancePerTenant` does not act like an instance-per-tenant but I have no clue why yet. I will try to add the specification tests now and see how that goes. Maybe it will help to find the problem.

Commit:
feat: simplify RequestServicesFeatureFactory since we will now always only have two parameters instead of one

alsami commented:
@tillig just to let you know I haven't forgotten about this - super busy at the moment, I will get back to it this week.

tillig commented:
No worries, take your time. I, too, am pretty swamped. :)

Commit:
feat: register service-scope-factory as instance-per-tenant and remove manual locking

Commit:
chore: rename ServiceScopeFactory delegate to ServiceScopeFactoryAccessor

Commit:
feat: add tests for M.E.D.I specifications

Commit:
chore: add test that proves ServiceScopeFactory is not resolved once per tenant

alsami commented:
> Just thought of something that might be interesting to try: set up the M.E.DI compatibility tests but with a tenant container.
> 
> That is...
> 
> * [Add a reference to the M.E.DI spec tests](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/c715e5f6c82f6481c299411fd0ec1f746f5f6aa5/test/Autofac.Extensions.DependencyInjection.Test/Autofac.Extensions.DependencyInjection.Test.csproj#L41)
> * [Create a `DependencyInjectionSpecificationTests` implementation](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/c715e5f6c82f6481c299411fd0ec1f746f5f6aa5/test/Autofac.Extensions.DependencyInjection.Test/Specification/BuilderSpecificationTests.cs#L9) where the way to build the `IServiceProvider` is:
>   
>   * Build the multitenant container.
>   * Set a "known tenant" so the tenant is always identified.
>   * Get the tenant's `IServiceProvider` and test against that.
> 
> That'd let us know if the logic here is going to hold. It may be interesting to have an internal static method that handles that little magic `AutofacChildLifetimeScopeServiceProviderFactory` registration so it can be used to wire that up in tests; or use the `AutofacMultitenantServiceProviderFactory` to build up the service provider so it's done the same as in the middleware.

So I have added the tests and chose the later for the setup. All the specification tests are running fine, which is misleading.

I have added another test-class `AssumedBehaviorTests` with one tests to prove that the `ServiceScopeFactoryAccessor` does not resolve the `IServiceScopeFactory` once per tenant. It's somehow acting like an `InstancePerDependency()` but I really do not understand why.

I've added this test as part of the second test-class

```csharp
[Fact]
public void SomeTest()
{
    var builder = new ContainerBuilder();
    builder.Register<ServiceScopeFactoryAccessor>(_ =>
    {
        return scope =>
        {
            var autofacChildLifetimeScopeServiceProviderFactory = new AutofacChildLifetimeScopeServiceProviderFactory(scope);
            var adapter = autofacChildLifetimeScopeServiceProviderFactory.CreateBuilder(new ServiceCollection());
            var serviceProvider = autofacChildLifetimeScopeServiceProviderFactory.CreateServiceProvider(adapter);
            return serviceProvider.GetRequiredService<IServiceScopeFactory>();
        };
    }).InstancePerTenant();

    var mtc = new MultitenantContainer(new StaticTenantIdentificationStrategy(), builder.Build());

    var serviceScopeFactoryAccessor = mtc.Resolve<ServiceScopeFactoryAccessor>();

    var firstServiceScopeFactory = serviceScopeFactoryAccessor.Invoke(mtc.GetCurrentTenantScope());
    var secondServiceScopeFactory = serviceScopeFactoryAccessor.Invoke(mtc.GetCurrentTenantScope());

    Assert.Same(firstServiceScopeFactory, secondServiceScopeFactory);
}
```

Even when it is isolated like it is here, the implementation-factory is executed twice, not once. Is this to be expected when the `StaticTenantIdentificationStrategy` always returns the same tenant-object?

tillig commented:
Just to acknowledge - I'm still here, I'm reading, but I'm not really _diving in and jamming on it_ because it sounds like you're just working through things. I'll keep lurking and pop in if there's something to add.

Commit:
feat: add new type to wrap IServiceScopeFactory that can be resolved once per tenant

alsami commented:
> Just to acknowledge - I'm still here, I'm reading, but I'm not really _diving in and jamming on it_ because it sounds like you're just working through things. I'll keep lurking and pop in if there's something to add.

I think I got it now. Instead of using a delegate, I created an adapter that takes an `IServiceScopeFactory` which is registered as `InstancePerTenant()`.

Once you've got time, please review the latest commit - I will try to adress the remaining bits regarding the package-references.

tillig commented:
Sorry, I somehow missed the notification about this most recent comment. I'll have to check out the branch to review, it's getting cumbersome not being able to navigate the project.

tillig commented on test/Autofac.Integration.AspNetCore.Multitenant.Test/Specification/AssumedBehaviorTests.cs:
```diff
@@ -0,0 +1,45 @@
+using System;
```
Missing file header.

tillig commented on test/Autofac.Integration.AspNetCore.Multitenant.Test/Specification/BuilderSpecificationTests.cs:
```diff
@@ -0,0 +1,38 @@
+using System;
```
Missing file header.

tillig reviewed (changes requested):
I started reviewing this, but I also found that there are some stale dependencies, like the analyzers and such, and I can't seem to comment on them. I think I will end up putting a PR in for your branch on your repo to help speed some of this along.

tillig committed:
Add file headers.

tillig committed:
README formatted for packaging.

tillig committed:
Updated analyzers, build, packaging for NET6.

Includes updates to better support VS Code - .NET TFM needs to be
ordered in descending order (VS Code uses the first one listed).

Most analyzers are built into the .NET 6 build, no need for potentially
conflicting analyzers in other TFMs.

Rulesets, .editorconfig, .csproj all brought into alignment with core
Autofac.

tillig committed:
Resolved new analyzer warnings.

tillig committed:
Updated dev dependencies.

tillig committed:
Enabled analyzers on sandbox projects.

tillig committed:
New client per integration test.

tillig committed:
Spelling words.

tillig committed:
Convert to file-scoped namespaces.

tillig committed:
Remove feature factory - only needed for net2.x.

tillig committed:
Semver => 6.0, publish .snupkg

tillig committed:
Spelling updates.

tillig committed:
Cast feature to IServiceProvidersFeature

Without the cast, it sets a `RequestServicesFeature` but that won't come
back later when things ask for an `IServiceProvidersFeature`. We need
the explicit cast.

tillig committed:
Verify factory adapter is InstancePerTenant.

tillig committed:
Tenant accessor must be transient

The TenantAccessor only identifies the tenant on construction, so if
it's a singleton it will cache the wrong tenant value for subsequent
requests.

tillig committed:
Remove 2.1-2.2 launch config.

Commit:
fix: re-add RequestServiceFeatureFactory to make Sandbox projects work

alsami commented:
@tillig I think we should be good to go now or is there anything left that needs to be done?

Commit:
chore remove ServiceFeature factory and dipose feature async

tillig commented:
I think we're good. I can get this merged in and released probably Monday.

tillig reviewed (approved):
🎉

alsami commented:
Yay! Glad we got this done, been a rough ride. Thanks for the support!

tillig commented:
Nice work! Here's hoping we don't find any more of these fun undocumented behaviors.

ghost commented:
There is a note at the bottom of this page (https://github.com/autofac/Autofac.Extensions.DependencyInjection/releases) which says it will be removed when issue fixed.  Is that considered fixed now, and the note should be removed?

tillig commented:
Oh, good point. Fixed!
